### ISSUE TITLE
More advanced automated tests

### ISSUE THREAD

**Issue by Beider** [DoubleDeez/MDFramework-Examples#8]
The automated tests should run all tests multiple times with different configurations. This should be done by having the clients disconnect and restarting the server before they reconnect.

In addtion we should also have some static nodes already existing in the scene that also perform some tests to test the current bug where these nodes are wiped from the replicator when the session ends.

**Comment by Beider** [comment]
Further addition to the tests are needed. To stay more in line with C# and to make the test more dynamic and organized we should use attributes instead of just invoking Test#.

Initial idea for this would be to do something like,
```csharp
[MDTest(Server)]
public void TestThatSomethingSomething() {}
public void TestThatSomethingSomethingValidate() {}

[MDTest(Client1)]
[MDTestSetting(Validate.Server, TestSomethingElseValidateServer)]
[MDTestSetting(Validate.Client2, TestSomethingElseValidateClient2)]
public void TestSomethingElse() {}
public void TestSomethingElseValidateServer() {}
public void TestSomethingElseValidateClient2() {}
```

However it is probably wise to take a hard look at the entire test framework to enhance it with more features. For instance it would be good to have an easy way to limit the tests you run in case you are working on one specific feature and don't want to run all tests.

In addition we should get automatic building on windows going so you only have to run the batch file instead of manually compiling inside godot first.

**Comment by Beider** [comment]
## General
Building on the concept above here is a more detailed concept for how testing should work. In short we will have Configs and Tests, each config will trigger a reconnect where it may do some setup (such as load new config files, completely reload all MDFramework classes, etc... Then it will reconnect. Then each test will be run for the config before the next config is loaded.

## MDTestManager
- Initialized by the loading scene that loads the tests
- Contains all the logic for loading configs / tests
- Contains networking code to start/ stop tests across the clients
- Contains logging methods for tests to log errors.
- Generally the brain of the automated tests.

## MDTestConfig

```csharp
[MDTestConfig(DevelopmentMode = false, TestGroup = "All", Repeat = 1)]
    [MDTestConfigBeforeConnect] // Invoked right before we are about to connect and start this config
    [MDTestConfigAfterConnect] // Invoked right after connecting for this config, before starting tests
    [MDTestConfigBeforeDisconnect] // Invoked when all tests are done right before disconnect
    [MDTestConfigAfterDisconnect] // Invoked after disconnect right before next test config will be started
```

Classes that contains setup for tests (could load configs, whatever). All tests are run once for each test config. All clients and server will reconnect for each test config. You can have as many before / after methods as you want.
- **DevelopmentMode**: If development mode is set to true only the configs with that flag set will be run.
- **TestGroup**: Set which group this test config should run, will only run tests in that group unless set to all
- **Repeat**: How many times to  repeat this test config, default 1

## MDTestClass + MDTest

```csharp
[MDTestClass(DevelopmentMode = false, TestGroup = "All", AlwaysRunInAllMode = false)]
    [MDTest(MDTestHost = MDTestPeer.Server, MDTestMode.Test)]

// All modes respects MDTestPeer to decide where to run
enum MDTestMode
{
    Test, // Normal test, executed as a test, looks for validation method after
    SetupClass, // Ran before all tests
    BeforeTest, // Run before every MDTestMode.Test
    AfterTest, // Run after every MDTestMode.Test
    ShutdownClass // Run after all tests are complete
}
enum MDTestPeer
{
    Server,
    Client1,
    Client2,
    All
}
```

A test class that contains tests. Each test is marked as an MDTest, by default we look for a method that is called `<same method name>Validate` for test validation. By default tests are run on server and validation on other clients. In the event that test host is set to Client1 or Client2 test will run on those clients and validation on others.
- **DevelopmentMode**: If development mode is set to true only the test classes tagged with development mode will be run
- **TestGroup**: Set which group this test belongs to, will only be run for test configs that tag for that group
- **AlwaysRunInAllMode**: If set to true this test will be run for configs that is set to TestGroup="All" even if this is in a different test group

## MDTestSetting

```csharp
[MDTestSetting]
    Settings.Validate / MDTestPeer
        Overwrites the validate methods, if specified will only run validation on the peers specified.

    Settings.TestOrder / int
        If set the test will respect this order, any test with a number will be done before all other tests in the class

    Setting.DevelopmentMode / bool true/false
        If set to true for any test then only tests with the setting DevelopmentMode will be run.
```

The test manager will be a child of the SceneRoot() directly like other autoloads. That means tests will now be able to load other scenes as well.


## General framework notes
- Framework should scan entire res:// for test classes
- No more need to make a scene for each test class or config class, we will instance.
- If you need a scene for your test you can instance that yourself from within test setup/shutdown
- Test framework will need more error handling surrounding tests to ensure that failing tests don't crash the framework
- The error handling should be on all levels, test config / test case / individual test.

## Nice to have in the future
- Profiling around every test / test class / test config
- Test framework could record nodes that are created and if a test doesn't cleanup after itself it should auto cleanup between test classes / test configs
- If we go for recording and cleaning up of nodes then we should ensure there is some way to tag nodes not to be cleaned up, maybe just register them with the test controller.
